File: Haskell/P7.hs
-------------------------------------------------------------------------------
-- Solved by Walsker
-- Problem 7: Flatten a nested list structure
-------------------------------------------------------------------------------

data NestedList x = Elem x | List [NestedList x]

flatten :: NestedList x -> [x]
flatten (Elem x) = [x]
flatten (List (h:t)) = (flatten h) ++ (flatten (List t))
flatten (List []) = []

File: Prolog/P7.pl
% -----------------------------------------------------------------------------
% Solved by Walsker
% Problem 7: Flatten a nested list structure
% -----------------------------------------------------------------------------

flatten([], []).
flatten([A|[]], B) :-
	is_list(A), flatten(A, B);
	B = [A].
flatten([A|B], X) :- 
	is_list(A), flatten(A, Y), flatten(B, Z), append(Y, Z, X);
	flatten(B, Y), X = [A|Y], !.